### PR TITLE
Groupby performance

### LINKED ISSUE #51: Graphing Performance is Slow
Graphing new time points runs at an incredibly slow pace -- about 3-5 seconds per update.  

The way that data is stored needs to be changed so that each frame update only deals with a small section of the dataframe, since profiling shows that the bottleneck is the serverside processing.  

This may come from creating an individual graph object for each datapoint or needing to process and filter the entire 10000 line dataframe for every new frame update, rather than a subsection.

### PR DESCRIPTION
**Pull request by alexanderkyim**
---
name: Group By Performance
about: Changes the structure of uploading data to be  

---

## Issues Addressed
Link to any issues this pull request addresses

## What has changed
Include each file that was changed, how it was changed (in broad terms), and why it was changed.
Example:
    tesla_ver/bubblechart.py:
        upload_data(): now saves the dataframe into a json dictionary with each time value as a key after grouping by time
        update_figure(): now only turns the correct year value into a dataframe, rather than the entire dataframe, and also 
                                    generates a single scatterplot per frame.
## Checklist of pre-requisites
All of these need to be checked in order to request a review on the Pull Request
-   [x] Does the code run?  
-   [x] Does the code follow the repository style?  
-   [x] Is the code tested?  
-   [x] Has Black been run on all the python files in the repository (with line length 120 `black -l 120 *`)?

## Pull Request Types
The following sections are specific to the three types of PRs. Enhancement, Bug Fix, and Design Change after determining the type of your PR delete the questions related to the other two.

# Bug Fix

## Describe or link to the issue that was fixed.
Graphing ran really slowly before the change, now it runs quickly and smoothly, closes #51 

## What is the new behavior?
It splits out the mdata and creates a dict of years to make it so that the entire dataframe only needs to be processed once

# Enhancement

## How to use the feature
`python wsgi.py`, and upload any file that matches `data/tesla_*.tsv`

### CODE REVIEW

File: tesla_ver/bubble_chart.py [hunk 1]
```diff
@@ -14,185 +19,198 @@ def generateBubbleChart(server):
...
-        df = pd.read_csv(fileish)
+        if None in [list_of_contents, list_of_filenames] or len(list_of_contents) != 1:
+            raise PreventUpdate
 
-        return df
+        def upload_string_to_df(content):
+            _, content_string = content.split(",")
+            decoded = base64.b64decode(content_string)
+            fileish = io.StringIO(decoded.decode("utf-8"))
+            return pd.read_csv(fileish, sep="\t")
```
Comment (DSWallach) [inline tesla_ver/bubble_chart.py:39]: I think we should split callbacks and non-callback functions into separate files. The non-callbacks should be in a library file that is installed with when running `setup.py`

File: tesla_ver/bubble_chart.py [hunk 2]
```diff
@@ -14,185 +19,198 @@ def generateBubbleChart(server):
...
+
+            contents is a string read from the upload component.
+            filename and date are just information paramteres
+            """
+            # Parses contents into dataframes
+            df_list = list()
+            filename_titles = list()
+            for content, filename in zip(contents, filenames):
+                df = upload_string_to_df(content)
+                # Separates time series into a single value
```
Comment (DSWallach) [inline tesla_ver/bubble_chart.py:52]: Expand on this

File: tesla_ver/bubble_chart.py [hunk 3]
```diff
@@ -14,185 +19,198 @@ def generateBubbleChart(server):
...
-        contents is a string read from the upload component filename and
-        date are just information paramteres
-        """
-        _, content_string = contents.split(",")
+        df = None
 
-        decoded = base64.b64decode(content_string)
-        fileish = io.StringIO(decoded.decode("utf-8"))
-        df = pd.read_csv(fileish)
+        if None in [list_of_contents, list_of_filenames] or len(list_of_contents) != 1:
```
Comment (DSWallach) [inline tesla_ver/bubble_chart.py:32]: Add a comment here describing the case this prevents

File: tesla_ver/bubble_chart.py (hunk 1 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py:42]: Just to follow up from our conversation, splitting out these functions into separate files will be addressed in a future branch.

File: tesla_ver/bubble_chart.py (hunk 2 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py:55]: addressed in 922d1c0

File: tesla_ver/bubble_chart.py (hunk 3 above)
Comment (alexanderkyim) [reply tesla_ver/bubble_chart.py:35]: addressed in 922d1c0

**Pull request merged**